File: 25. Reverse Nodes in k-Group.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        
        ListNode prev=dummy, curr=head, fast=head, next=null;
        ListNode con=dummy, tail=dummy;
        while (true) {
            int count = 0;
            while (fast != null && count < k) {
                fast = fast.next;
                count++;
            }
            if (count != k) {
                return dummy.next;
            }
            
            con = tail;
            tail = curr;
            while (count > 0) {
                next = curr.next;
                curr.next = prev;
                prev = curr;
                curr = next;
                count--;
            }
            con.next = prev;
            tail.next = curr;
        }
    }
}

File: 25. Reverse Nodes in k-Group.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        dummy = ListNode(0)
        dummy.next = head
        fast, slow = head, head
        prev_k = dummy
        while True:
            count = 0
            while fast and count < k:
                fast = fast.next
                count += 1
            if count == k:
                pre, cur = fast, slow
                for _ in range(k):
                    cur.next, cur, pre = pre, cur.next, cur 
                prev_k.next, prev_k, slow = pre, slow, fast
            else:
                return dummy.next